### DISCUSSION TITLE
Any chance this could become a prettier plugin?

### DISCUSSION THREAD

**Discussion by mandrasch** [Shifu462/twiggy#24] [Q&A]
Hi, just stumbled over twiggy. I code with Craft CMS, great effort to maintain something new! Thanks! 👏 

As far as I understand twiggy is VS Code only, right?

Is there any chance the language server could be somehow used for a prettier plugin?

It's still a big painpoint for me that there is no way to auto-format in PhpStorm / WebStorm and other IDEs consistently - in addition to VS code. 

I'm not experienced with language servers - just read about Astro switchting to Volar als language server tooling. https://astro.build/blog/astro-and-volar/ 

And I heard that a language server is the first step to get a prettier plugin up and running. Therefore I'm trying to figure out how much effort is needed to get this done for twig.

Thanks very much for any hints! 

Regards, Matthias

**Comment by moetelo** [comment]
Thank you too!

> twiggy is VS Code only, right?

Also works in [Neovim](https://github.com/moetelo/twiggy/issues/12) and [Sublime](https://github.com/moetelo/twiggy/issues/7)

> prettier

This could work: https://github.com/zackad/prettier-plugin-twig-melody \
Untested.

The main discussion about formatting is here: https://github.com/moetelo/twiggy/issues/4

**Reply by mandrasch** [reply]
Oh thanks! 🙏  I knew twig-melody plugin but it was unmaintained for some time and did not properly support some Craft CMS specialties. Didn't know there is an active fork now.

**Reply by moetelo** [reply]
Please ping me if it works for you :)

**Reply by mandrasch** [reply]
The original plugin had a lot of open issues https://github.com/trivago/prettier-plugin-twig-melody/issues, especially the fix for https://github.com/trivago/prettier-plugin-twig-melody/issues/47 would be important for using it in Craft CMS. Bit of special markup of course, not official twig. 

So in general - I guess there is some (community) work needed to fix the important issues. I'll try to setup an online playground (https://github.com/zackad/prettier-plugin-twig-melody/discussions/29) for testing it more easily. 

But having a new fork with someone taking the lead is new hope :))

**Reply by webrgp** [reply]
@mandrasch I've been exploring ways to port @kaermorchen work on [twig-language-server](https://github.com/kaermorchen/twig-language-server) into a prettier plugin, which is how I end up finding Twiggy.

And while reviewing the issues, I've ran into [Twig CS Fixer](https://github.com/vincentlanglet/twig-cs-fixer/) which worked great with Craft! View [this issue](https://github.com/VincentLanglet/Twig-CS-Fixer/issues/146) for more details, or hit me up if you run into issues. It's about time DX for Twig get some love.

**Reply by moetelo** [reply]
Twig CS Fixer is great, although it fixes only Twig-related issues

I'll think about integrating it into Twiggy

**Reply by mandrasch** [reply]
@webrgp thanks for information! 👍 

My dream: Auto-formatting in Visual Studio Code as well as in PhpStorm. ✨ (or other IDEs)

Prettier seems to be the best option since it's already integrated in PhpStorm out of the box. Just a robust and maintained prettier plugin for twig is the missing puzzle piece.

**Reply by webrgp** [reply]
@mandrasch Like I've mentioned, I've started out with the same dream, so I will try to save you some time by sharing my own journey: I wanted something that could fix formatting both in the editor and as a "pre-commit" action.

I thought that a prettier plugin would be the solution, and it makes sense since you can run prettier both in the editor and in the command line. But I've been discovering that the inherited issue with Twig is that it can be implemented in different ways and be extended to fit the larger system needs (Craft vs Symphony for example).

Twig Melody tried to tackle that issue by allowing the user to [add additional multi-tags](https://github.com/trivago/prettier-plugin-twig-melody?tab=readme-ov-file#twigmultitags-default-), but that has a ceiling and is limited to pairs. That's because it's parser is written in javascript. And don't even get me started on [type hinting](https://github.com/moetelo/twiggy/issues/14) on [twig](https://github.com/twigphp/Twig/pull/4009#issuecomment-2002536778)!

Twig CS Fixer on the other hand is written in PHP, which allow use to hook into the twig implementation of the host framework, where we can add any additional functionality such as functions, tags, filters and even the ones added through custom modules and plugins. Even this VS Code extension has some PHP baked in.

So I've come to realize that the path to achieve robust twig DX experience is to have some PHP baked in, which why I am excited about this extension and what @moetelo has done so far. I hope this would save you some time, but if you do find other working implementations, please do share!

**Reply by mandrasch** [reply]
@webrgp Thanks very much for sharing your journey and the information, interesting! 👍👍 

"Baked in PHP" can also a bit of a struggle, because normally you want composer to be run with your projects php version, e.g. for Docker, DDEV, Lando, Sail etc. - therefore I liked the way of how prettier works with plugins in node_modules/ and no extra config needed. 

But I saw that for Twiggy it is possible to define `twiggy.phpExecutable`.  

Let's say prettier is not to way to go (as you experienced so far)... 

Then the missing puzzle piece is "just" a Twiggy-port for PhpStorm to get auto-formatting working in both IDEs? 🤔

**Reply by webrgp** [reply]
PHPStorm relies on PHP, [either through its included AMP (Apache, MySQL, PHP) or through a local/remote environment](https://www.jetbrains.com/help/phpstorm/configuring-php-development-environment.html). Even when using the [Symfony Support](https://github.com/Haehnchen/idea-php-symfony2-plugin) plugin, which is written in Java, it also [uses PHP](https://github.com/Haehnchen/idea-php-symfony2-plugin/tree/63c305c7edebef70dae77e5b90171efcf21ccc23/src/main/resources/fileTemplates) to access it.

So yes, I think any robust twig formatter would have to have some PHP component to it. As of now, I think this can be achieve using [Twig CS Fixer](https://github.com/VincentLanglet/Twig-CS-Fixer) because it allows us to [configure additional rules](https://github.com/VincentLanglet/Twig-CS-Fixer/blob/main/docs/rules.md#configurable-rules) by extending [AbstractFixableRule](https://github.com/VincentLanglet/Twig-CS-Fixer/blob/main/src/Rules/AbstractFixableRule.php).

**Reply by webrgp** [reply]
And let me reinforce that this is where I've land on it at this time, until I can either find or be shown a better way :wink:

**Reply by mandrasch** [reply]
Yes agreed, just wanted to make the point that the question of php executable settings wouldn't be needed if prettier is used https://www.jetbrains.com/help/phpstorm/prettier.html (And people use prettier anyway for css and js via node_modules/)

Funny sidenote: PhpStorm has it's own Twig formatting integration since 2013 out of the box. So a shared configuration standard for this and Twiggy in VSCode could also ease some pain when the same settings are used (in order to avoid git diff chaos when devs auto-format in different IDEs). But this is also complicated I guess. Or a twiggy-port for PhpStorm. 🤔 

Just for clarification: Twig CS Fixer currently has no auto-format/auto-indent features for twig/html, right?

**Reply by webrgp** [reply]
Correct! I've cloned [PHP CS Fixer for Visual Studio Code](https://github.com/junstyle/vscode-php-cs-fixer), which basically bridge the same gap for [PHP CS Fixer](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer), to create an extension for the Twig CS Fixer in the same mold. 

In the mean time, I've bookmarked [Run On Save](https://marketplace.visualstudio.com/items?itemName=emeraldwalk.RunOnSave) extension to try and see if that will work for now.

**Reply by moetelo** [reply]
To be clear, it would be great if we integrated Twig CS Fixer into Twiggy. That would already happen if I wasn't that focused on completion and typing features.

So, if you have time for this, then feel free to send a PR. You can also ask me for any help with this.
